File: Top_100_Liked_Questions/438._Find_All_Anagrams_in_a_String/solution.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
    let len_s = s.length;
    if (len_s === 0) {
        return [];
    }
    
    let len_p = p.length;
    if (len_p === 0) {
        return [];
    }
    
    if (len_p > len_s) {
        return [];
    }
    
    let p_cnt = new Array(26).fill(0);
    let curr_cnt = new Array(26).fill(0);
    let ascii_a = 'a'.charCodeAt(0);
    for (let i=0; i<len_p; i++) {
        p_cnt[p.charCodeAt(i) - ascii_a]++;
        curr_cnt[s.charCodeAt(i) - ascii_a]++;
    }
    
    let ans = p_cnt.toString() != curr_cnt.toString() ? [] : [0];
    for (let i=0; i<len_s-len_p; i++) {
        if (s[i] != s[i+len_p]) {
            curr_cnt[s.charCodeAt(i) - ascii_a]--;
            curr_cnt[s.charCodeAt(i+len_p) - ascii_a]++;
        }
        
        if (curr_cnt.toString() === p_cnt.toString()) {
            ans.push(i+1);
        }
    }
    
    return ans;
};

File: Top_100_Liked_Questions/438._Find_All_Anagrams_in_a_String/solution.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        # sliding window method
        if not s or not p:
            return []
        
        len_p = len(p)
        if len_p > len(s):
            return []
        
        ord_a = ord('a')
        
        cnt_p = [0] * 26
        curr_cnt = [0] * 26
        
        for i in range(len_p):
            cnt_p[ord(p[i]) - ord_a] += 1
            curr_cnt[ord(s[i]) - ord_a] += 1
        
        ans = []
        if curr_cnt == cnt_p:
            ans.append(0)
            
        for i in range(len(s)-len_p):
            if s[i] != s[i+len_p]:
                curr_cnt[ord(s[i]) - ord_a] -= 1
                curr_cnt[ord(s[i+len_p]) - ord_a] += 1
            
            if curr_cnt == cnt_p:
                ans.append(i+1)
        
        return ans